### DISCUSSION TITLE
Code refactoring in Zed

### DISCUSSION THREAD

**Discussion by hajarrashidi** [zed-industries/zed#8623] [Feature Requests]
One of my favorite aspects of JetBrains IDEs is their robust refactoring functionality. The ability to search for references when changing filenames, class names, configurations, or variables is invaluable. This is a major reason I prefer JetBrains over VS Code. Their search tools for finding text and files are also excellent, but that's a separate discussion.

I've reviewed the Zed roadmap (https://zed.dev/roadmap) and don't see any specific plans to enhance refactoring capabilities. While I generally agree with the 'less is more' philosophy, I believe refactoring tools are a core IDE feature. It would be ideal if either plugin developers or the Zed core team could implement these improvements.  What are your thoughts on this?
<img width="801" alt="Screenshot 2024-02-29 at 17 33 46" src="https://github.com/zed-industries/zed/assets/4338882/dfe06f41-8890-46ab-a932-c07a2f56b0c0">

**Comment by Moshyfawn** [comment]
Although a completely different workflow, Zed provides a versatile tool called [multi-buffers](https://zed.dev/features#multi-buffers) that allows you to perform various project-wide operations, including refactoring. There's an official Zed [demo](https://youtu.be/egI75sLT164?si=kkd_VNBL0U2Xa4iz) where Antonio shows how he goes about refactoring a Rust struct usage across his project.

**Comment by pablo2909** [comment]
+1 on this, that is the main reason for me not switching to Zed :)

**Comment by kurtbuilds** [comment]
The partial solution above, multi-buffers, don't address the full power of refactors in IntelliJ.

In IntelliJ, there's a "move to another file" refactor, which moves an item (struct, fn, enum) to another file, automatically updating all the imports in the process.

This is a huge feature that's missing for me and prevents me from switching to Zed full time.

**Comment by nnra6864** [comment]
I 100% agree that refactoring would be a huge thing for Zed.
I am still trying it out and that seems to be the most major thing I am missing from other IDEs tbh.
Being able to just change the c# namespace or variable name with a few clicks is huge.

**Comment by jakubkaczor** [comment]
The problem with implementing code refactorings is that it would need to be implemented for each language and this requires a lot of work and polylingual expertise. The LSP protocol only specifies simple refactorings that do not require much user input or user interface, so language servers and code actions cannot be leveraged.

Did the developers of Zed considered inventing their own refactoring protocol? It is too early to judge with any confidence, but I suspect the authors can be in somewhat suitable position to, at least, think about it for the future. For such protocol to gain popularity, the code editor itself would have to become popular, so that language server authors would have interest in implementing it. Out of the most popular code editors[^1]:
- Microsoft's Visual Studio Code did not come up with such protocol, neither with an LSP extension for years[^2],
- JetBrains probably has a conflict of interests,
- terminal editors possibly have a limited user base excluding beginner developers looking for a GUI, out of the box, and friendly experience.

Atom used to be the most popular[^3], not so far mentioned, open-source editor. If Zed is in the same spirit, it has a potential to gain traction and help such protocol become widely adopted. Some details about attempts to work around the limited protocol can probably be learned from the creators of [Wrangler](https://refactoringtools.github.io/docs/wrangler/)[^4].

[^1]: https://survey.stackoverflow.co/2024/technology#2-integrated-development-environment
[^2]: https://github.com/microsoft/language-server-protocol/issues/1164
[^3]: https://survey.stackoverflow.co/2021#technology-most-popular-technologies
[^4]: https://www.youtube.com/watch?v=grQP1ikomek

**Comment by danballance** [comment]
Not being able to `Refactor -> Move` code in Zed, the same way I can in PyCharm, is a significant problem for me. It's such a core part of my workflow and something I can do so quickly in PyCharm.

A common use case is where an LLM has completed some code. It's functionally working and an okay first draft - but now I need to clean it up, write tests and make it my own. The first thing I'll do is make some different decisions about the layout of the code and move various classes to different modules. This is super-fast and easy to do in PyCharm - it's a very manual and error-prone process in Zed. So I have to fire PyCharm up again :/

**Comment by Havunen** [comment]
Anybody interested adding refactor functionality for Rust language?

**Comment by danballance** [comment]
This is what I need. Python and Rust. For now, I've returned to PyCharm and Rust Rover.

Sent from [Proton Mail](https://proton.me/mail/home) for Android.

-------- Original Message --------

**Comment by David-OConnor** [comment]
Trip report: Tried using Zed on my tablet (Rust in this case, but this functionality affects multiple languages), (JetBrains is a pig...; the CPU and battery on this device do not like it!) and I immediately ran into this.
